std/compress/io.go

```diff
@@ -143,3 +143,99 @@ func ShiftLeft(api frontend.API, slice []frontend.Variable, shiftAmount frontend
 	}
 	return res
 }
+
+// NumReader takes a sequence of words [ b₀ b₁ ... ], along with a base r and length n
+// and returns the numbers (b₀ b₁ ... bₙ₋₁)ᵣ, (b₁ b₂ ... bₙ)ᵣ, ... upon successive calls to Next()
+type NumReader struct {
+	api         frontend.API
+	toRead      []frontend.Variable
+	radix       frontend.Variable
+	numBound    frontend.Variable
+	wordsPerNum int
+	last        frontend.Variable
+}
+
+// NewNumReader returns a new NumReader
+// toRead is the slice of words to read from
+// numNbBits defines the radix as r = 2ⁿᵘᵐᴺᵇᴮⁱᵗˢ (or rather numNbBits = log₂(r) )
+// wordNbBits defines the number of bits in each word such that n = numNbBits/wordNbBits
+// it is the caller's responsibility to check 0 ≤ bᵢ < r ∀ i
+func NewNumReader(api frontend.API, toRead []frontend.Variable, numNbBits, wordNbBits int) *NumReader {
+	wordsPerNum := numNbBits / wordNbBits
+
+	if wordsPerNum*wordNbBits != numNbBits {
+		panic("wordNbBits must be a divisor of 8")
+	}
+
+	return &NumReader{
+		api:         api,
+		toRead:      toRead,
+		radix:       twoPow(api, wordNbBits),
+		numBound:    twoPow(api, numNbBits),
+		wordsPerNum: wordsPerNum,
+	}
+}
+
+func twoPow(api frontend.API, n int) frontend.Variable {
+	res := frontend.Variable(1)
```

Comment:
working with a constant here; I would just use `1 << n` or if it's larger than 64bits, bit.Int Rsh method, no need to use the API and incur the extra frontend compile cost

Reply:
Encountered an `n=248` in the zkevm monorepo.

Reply:
then just big.Int LSH 👍